File: app/src/main/java/com/dumuzeyn/mp3player/HomeContent.java
package com.dumuzeyn.mp3player;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class HomeContent {
    final List<Track> recentlyPlayed;
    final List<Track> recentlyAdded;
    final List<Track> mostPlayed;
    final List<Track> favorites;
    final List<Track> allFavorites;
    final List<Playlist> playlists;
    final List<String> artists;
    final List<String> albums;
    final Map<String, ArrayList<Track>> folders;
    final Map<String, ArrayList<Track>> artistTracks;
    final Map<String, ArrayList<Track>> albumTracks;
    final Map<String, ArrayList<Track>> genreTracks;

    HomeContent(List<Track> recentlyPlayed, List<Track> recentlyAdded,
            List<Track> mostPlayed, List<Track> favorites, List<Track> allFavorites,
            List<Playlist> playlists,
            List<String> artists, List<String> albums,
            Map<String, ArrayList<Track>> folders,
            Map<String, ArrayList<Track>> artistTracks,
            Map<String, ArrayList<Track>> albumTracks,
            Map<String, ArrayList<Track>> genreTracks) {
        this.recentlyPlayed = immutable(recentlyPlayed);
        this.recentlyAdded = immutable(recentlyAdded);
        this.mostPlayed = immutable(mostPlayed);
        this.favorites = immutable(favorites);
        this.allFavorites = immutable(allFavorites);
        this.playlists = Collections.unmodifiableList(new ArrayList<>(playlists));
        this.artists = immutable(artists);
        this.albums = immutable(albums);
        this.folders = immutableGroups(folders);
        this.artistTracks = immutableGroups(artistTracks);
        this.albumTracks = immutableGroups(albumTracks);
        this.genreTracks = immutableGroups(genreTracks);
    }

    static HomeContent empty() {
        return new HomeContent(Collections.emptyList(), Collections.emptyList(),
                Collections.emptyList(), Collections.emptyList(), Collections.emptyList(),
                Collections.emptyList(), Collections.emptyList(), Collections.emptyList(),
                Collections.emptyMap(), Collections.emptyMap(), Collections.emptyMap(),
                Collections.emptyMap());
    }

    private static <T> List<T> immutable(List<T> source) {
        return Collections.unmodifiableList(new ArrayList<>(source));
    }

    private static Map<String, ArrayList<Track>> immutableGroups(
            Map<String, ArrayList<Track>> source) {
        LinkedHashMap<String, ArrayList<Track>> copy = new LinkedHashMap<>();
        for (Map.Entry<String, ArrayList<Track>> entry : source.entrySet()) {
            copy.put(entry.getKey(), new ArrayList<>(entry.getValue()));
        }
        return Collections.unmodifiableMap(copy);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/HomeContent.kt
package com.dumuzeyn.mp3player

import java.util.Collections
import java.util.LinkedHashMap

class HomeContent(
    recentlyPlayed: List<Track>,
    recentlyAdded: List<Track>,
    mostPlayed: List<Track>,
    favorites: List<Track>,
    allFavorites: List<Track>,
    playlists: List<Playlist>,
    artists: List<String>,
    albums: List<String>,
    folders: Map<String, ArrayList<Track>>,
    artistTracks: Map<String, ArrayList<Track>>,
    albumTracks: Map<String, ArrayList<Track>>,
    genreTracks: Map<String, ArrayList<Track>>,
) {
    @JvmField val recentlyPlayed = immutable(recentlyPlayed)
    @JvmField val recentlyAdded = immutable(recentlyAdded)
    @JvmField val mostPlayed = immutable(mostPlayed)
    @JvmField val favorites = immutable(favorites)
    @JvmField val allFavorites = immutable(allFavorites)
    @JvmField val playlists: List<Playlist> = Collections.unmodifiableList(ArrayList(playlists))
    @JvmField val artists = immutable(artists)
    @JvmField val albums = immutable(albums)
    @JvmField val folders = immutableGroups(folders)
    @JvmField val artistTracks = immutableGroups(artistTracks)
    @JvmField val albumTracks = immutableGroups(albumTracks)
    @JvmField val genreTracks = immutableGroups(genreTracks)

    companion object {
        @JvmStatic
        fun empty(): HomeContent = HomeContent(
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList(),
            emptyMap(),
            emptyMap(),
            emptyMap(),
            emptyMap(),
        )

        private fun <T> immutable(source: List<T>): List<T> =
            Collections.unmodifiableList(ArrayList(source))

        private fun immutableGroups(
            source: Map<String, ArrayList<Track>>,
        ): Map<String, ArrayList<Track>> {
            val copy = LinkedHashMap<String, ArrayList<Track>>()
            source.forEach { (name, tracks) -> copy[name] = ArrayList(tracks) }
            return Collections.unmodifiableMap(copy)
        }
    }
}
